runtime/compiler/wasm/gen/main.go

```diff
@@ -0,0 +1,355 @@
+
+	count := len(ins.Opcodes)
+
+	// prefix
+	for i := 0; i < count-1; i++ {
+		if i > 0 {
+			b.WriteString(", ")
+		}
+		opcode := ins.Opcodes[i]
+		_, err := fmt.Fprintf(&b, "0x%x", opcode)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// final opcode
+	if count > 1 {
+		b.WriteString(", ")
+	}
+	_, err := b.WriteString(ins.OpcodeIdentifier())
+	if err != nil {
+		panic(err)
+	}
+
+	return b.String()
+}
+
+func (ins instruction) Opcode() byte {
+	return ins.Opcodes[len(ins.Opcodes)-1]
+}
+
+func (ins instruction) OpcodeIdentifier() string {
+	return fmt.Sprintf("opcode%s", ins.Identifier())
+}
+
+type instructionGroup struct {
+	Depth        int
+	Instructions []instruction
+}
+
+func (group instructionGroup) GroupByOpcode() map[string]instructionGroup {
+	result := map[string]instructionGroup{}
+
+	for _, ins := range group.Instructions {
+		innerDepth := group.Depth + 1
+		atEnd := len(ins.Opcodes) <= innerDepth
+		opcode := ins.Opcodes[group.Depth]
+		var key string
+		if atEnd {
+			key = ins.OpcodeIdentifier()
+		} else {
+			key = fmt.Sprintf("0x%x", opcode)
+		}
+		innerGroup := result[key]
+		innerGroup.Depth = innerDepth
+		innerGroup.Instructions = append(innerGroup.Instructions, ins)
+		result[key] = innerGroup
+	}
+
+	return result
+}
+
+var trailingWhitespaceRegexp = regexp.MustCompile("(?m:[ \t]+$)")
+
+const target = "instructions.go"
+
+func main() {
+
+	f, err := os.Create(target)
+	if err != nil {
+		panic(fmt.Errorf("could not create %s: %w\n", target, err))
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	var generateSwitch func(group instructionGroup) (string, error)
+
+	templateFuncs := map[string]interface{}{
+		"goGenerateComment": func() string {
+			// NOTE: must be templated/injected, as otherwise
+			// it will be detected itself as a go generate invocation itself
+			return "//go:generate go run ./gen/main.go"
+		},
+		"switch": func(group instructionGroup) (string, error) {
+			res, err := generateSwitch(group)
+			if err != nil {
+				return "", err
+			}
+			pad := strings.Repeat("\t", group.Depth+1)
+			padded := pad + strings.Replace(res, "\n", "\n"+pad, -1)
+			trimmed := trailingWhitespaceRegexp.ReplaceAll([]byte(padded), nil)
+			return string(trimmed), nil
+		},
+	}
+
+	parsedSwitchTemplate := template.Must(
+		template.New("switch").
+			Funcs(templateFuncs).
+			Parse(switchTemplate),
+	)
+
+	parsedFileTemplate := template.Must(
+		template.New("instructions").
+			Funcs(templateFuncs).
+			Parse(fileTemplate),
+	)
+
+	generateSwitch = func(instructions instructionGroup) (string, error) {
+		var b strings.Builder
+		err := parsedSwitchTemplate.Execute(&b, instructions.GroupByOpcode())
+		if err != nil {
+			return "", err
+		}
+		return b.String(), nil
+	}
+
+	declare := func(instructions []instruction) {
+		err = parsedFileTemplate.Execute(f,
+			instructionGroup{
+				Depth:        0,
+				Instructions: instructions,
+			},
+		)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	declare([]instruction{
```

Comment:
The main contribution of this PR: decoratively specify the instructions and generate the code for the types, the writing, the reading, and the dispatch/switch for them (including support for opcodes with prefixes)